sdk/tests/conformance2/buffers/getBufferSubDataAsync.html

```diff
@@ -0,0 +1,249 @@
+  debug("");
+  debug("Test that getBufferSubDataAsync successfully works with dstOffset and length");
+
+  clearDstData();
+  wtu.shouldGenerateGLError(gl, gl.NO_ERROR,
+      "promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, 0, dstData, 2, 2)");
+  return promise.then(function(arr) {
+    resolvedArray = arr;
+    shouldBeTrue("areArraysEqual(resolvedArray.slice(0, 2), [0, 0])");
+    shouldBeTrue("areArraysEqual(resolvedArray.slice(2, 4), floatArray.slice(0, 2))");
+    shouldBeTrue("areArraysEqual(resolvedArray.slice(4), [0, 0, 0, 0, 0])");
+  });
+});
+
+allPromises = allPromises.then(function() {
+  clearDstData();
+  wtu.shouldGenerateGLError(gl, gl.NO_ERROR,
+      "promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, 0, dstData, dstData.length, 0)");
+  return promise.then(function(arr) {
+    resolvedArray = arr;
+    shouldBeTrue("areArraysEqual(resolvedArray, [0, 0, 0, 0, 0, 0, 0, 0, 0])");
+  });
+});
+
+allPromises = allPromises.then(function() {
+  debug("Test that getBufferSubDataAsync fails when given a dstOffset+length beyond the end of dstData");
+
+  var code = "promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, 0, dstData, dstData.length - 1, 2)";
+  wtu.shouldGenerateGLError(gl, gl.INVALID_VALUE, code);
+  return expectRejected(promise, code);
+});
+
+allPromises = allPromises.then(function() {
+  debug("");
+  debug("Test that getBufferSubDataAsync fails when given a buffer with its size larger than the original data");
+  var code = "promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, 0, extraLargeBuffer)";
+  wtu.shouldGenerateGLError(gl, gl.INVALID_VALUE, code);
+  return expectRejected(promise, code);
+});
+
+allPromises = allPromises.then(function() {
+  debug("Test that getBufferSubDataAsync fails when offset summed with buffer length is larger than the size of the original data size");
+  var code = "promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, dstData.byteLength + 1, dstData)";
+  wtu.shouldGenerateGLError(gl, gl.INVALID_VALUE, code);
+  return expectRejected(promise, code);
+});
+
+allPromises = allPromises.then(function() {
+  var code = "promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, 1, dstData)";
+  wtu.shouldGenerateGLError(gl, gl.INVALID_VALUE, code);
+  return expectRejected(promise, code);
+});
+
+allPromises = allPromises.then(function() {
+  debug("Test that getBufferSubDataAsync fails when 0 is bound to the target");
+  var code = "promise = gl.getBufferSubDataAsync(gl.ELEMENT_ARRAY_BUFFER, 0, dstData)";
+  gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, null);
+  wtu.shouldGenerateGLError(gl, gl.INVALID_OPERATION, code);
+  return expectRejected(promise, code);
+});
+
+allPromises = allPromises.then(function() {
+  debug("Test that getBufferSubDataAsync fails when offset is less than 0");
+  var code = "promise = gl.getBufferSubDataAsync(gl.ARRAY_BUFFER, -1, dstData)";
+  wtu.shouldGenerateGLError(gl, gl.INVALID_VALUE, code);
+  return expectRejected(promise, code);
+});
+
```

Comment:
A couple more tests that might be useful:
 - Check that getBufferSubDataAsync on an uninitialized (gl.bufferData with null) buffer returns zeros.
 - Check that getBufferSubDataAsync immediately followed by a buffer resize to be too small doesn't crash.